[Draft] Consume control node capacity

kdelee opened the pull request:
I've tested with container groups in a local minikube cluster, but it generally results in desired behavior
of only starting container group jobs when we have a control node with
capacity to run them. 

Read @beeankha 's comment below to explore what remains to implement the same thing completely for all jobs that have a separate control node.

<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "Consider control node capacity when selecting a controller node, and consume capacity for controller node"
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This patch addresses https://github.com/ansible/awx/issues/10694

For container groups in particular, there is currently no controller-side mechanism to throttle how many jobs get launched. This can lead to overwhelmed control nodes, and control of jobs unevenly spread across cluster (e.g., you have 3 control pods but 1 sits idle while the other two are overtaxed).

Even before we start to see bad things like control pod eviction from resource consumption, things start to happen like jobs sitting in `'waiting'` for over a min, which can lead to the jobs getting "reaped" but then running and succeeding anyway! By counting capacity consumption for controlling jobs, we at least have the "capacity slider" as a tool to have _some_ way to manage how many jobs a control pod will be controlling/proccesing events for at any given time.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

TODO:
  - [ ] add some tests
  - [ ] do some exploratory testing in a production k8s deployment
  - [ ] decide if we want to fully implement same thing for all jobs w/ separate control node in this PR, or in a follow up PR

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

kdelee committed:
Consume control node capacity

This is a rough start, I've only tested with container groups
Needs several refactors, but it generally results in desired behavior
of only starting container group jobs when we have a control node with
capacity to run them. May not be too difficult to extend to other job
types that have a seperate control node.

rebeccahhh committed:
change the query for available capacity

kdelee committed:
Add staticmethod to find best control instance

Maybe we can use this elsewhere next

rebeccahhh committed:
allow project & inventory updates to run on control nodes, also edit consume capacity debug message to see what each job gets in terms of capacity

rebeccahhh committed:
fix django query so it excludes successful jobs, because we don't care about them

kdelee committed:
use hostname for control node, not whole object

kdelee committed:
WELL...lets try and keep track ourselves

PS this is not valid python please fix

rebeccahhh committed:
fixed some python syntax, but found some logical errors and started working through those (or maybe my brain is just struggling)

kdelee committed:
Cut out extra loop over instances

We already loop over instances a few times, this cuts out one more and
builds the map of instance:remaining_capacity in an existing loop.

Delete a old key instances_remaining_capacity in the graph, because it
is not being used anywhere. I think this indicates someone in the past
had a similar intention as we have here for keeping track of the
remaining capacity on individual nodes, not just the group.

kdelee committed:
not all Jobs have the controller_node attribute

kdelee committed:
don't reap WF jobs when looking for orphans

workflow jobs don't have a control or execution node, but
they are not "orphaned" in the sense we are worried about in this
clause.

rebeccahhh committed:
adding a warning for when an instance group is not found for a control node, also modified if/else to be more specific around node_type

rebeccahhh committed:
when settings.IS_K8s is true this code will never be reached, rendering it useless.

rebeccahhh committed:
add a comment, hopefully interactive rebase this message out later

kdelee committed:
clean up inefficient debugging in capacity method

kdelee committed:
Refactor two identical loops to one

We were looping over the instances and excluding any disabled ones in
two loops, one after another. Just do the work in the one loop.

kdelee committed:
hype up this function

Not yet used anywhere, but its cooL!

kdelee committed:
use complete words as variable names

rebeccahhh commented:
**Potential problems with our code:**
- awx/main/managers.py::250 - we saw some situations where the control node was not found in the mapping of the nodes to instance group. 
    * It must have appeared _after_ the mapping was made, this means we have more capacity than we originally planned for.
    * Consequence is that we will not attribute consumed capacity to the instance group that it must belong to.
    * We added a warning log, we want to test in production k8s deploy to see if we can make this occur in a prod environment as opposed to a dev env.

**How we think it works:**
1. task manager inits and gets the lock
2. the graph is an empty dict - then on line 89 of task_managers.py in the post init the graph is initialized with keys that are the names of insatnce groups and values that are dummy values (i.e. 0)
3. On line 104 it seems like we annotate their total potential capacity (nothing consumed so far)
4. We see the graph next in process_tasks on line 187 where it is implicitly passed in calculate_capacity_consumed to capacity_values from manager.py along with the list of all running and waiting task_managers.py
5. in capacity_values we zero out the graph. Why did we even do `#3` if we were just going to do this again.
6. then we loop over the tasks and decide how much capacity is being consumed in each instance group by the tasks and update the graph (lines 267 and 293).
    * originally we were only tracking the task impact on the execution instance group, but now we track the task impact on the instance groups that the control node belongs to too.
    * our changes will affect all jobs - at the beginning of the task maanger
7. For jobs that are using get_and_consume_capacity_on_control_node_with_sufficient_capacity to select and record the task_impact for the their control node we update the graph.
8. container group jobs will only leave pending once they have capacity on a control node, otherwise they will stay in pending.
9. for all non-container group jobs on line 568 in task_manager.py we finally check the graph by calling get_remaining_capacity in which we consult the graph (line 677).


**Remaining questions:**
- what's up with `#5`?? why do we initialize and then zero out the graph twice.
- A risk/to be figured out: we know the task manager has problems when there are too many jobs to process, we are intentionally putting a bottleneck on jobs leaving pending and will be collected and looped over by the task manger and may stress the start_task_limit

**Things we claim this PR does:**
- With these changes container group jobs fully consider and record their task impact on control nodes.
- All other nodes are impacted by our changes when the graph is initially built but still don't consider their impact on control nodes when transitioning from pending into waiting.
- If we wanted to fully apply consideration of capacity for selection of control node for _non-container_ group jobs we need to select the control node at some point between line 568 and line 597 (i.e. during process_pending_tasks)

kdelee committed:
fix typo causing failures

kdelee committed:
merge three nearly identica data structures

instances_by_hostname
instances_partial
self.real_instances

Were all...mappings of instance host names to some form or another of
data about the instance. Instead of having three of these, lets just get
by with one.

kdelee committed:
find most available node with heap

This is a re-implementation using some wrapper objects and the heapq module.
This may not actually be much faster or even easier to understand :D

But the end result is the same, we get the control node with
most available capacity. Going to shop this version around to others
and see which one is preferred.

kdelee committed:
Give map of instances a more dignified name

Move the map of instances in the task manager into its own class. Make
the init of the graph using that data a method of that class. Move these
classes that help encapsulate the capacity and ability of all the instance the task
manager will be assigning work to into another file. It would be great
to add some tests of these classes themselves now that they are discreet
units.

kdelee committed:
Give the two graphs distinctive names

One is a graph of instance groups and the tracking data for what
capacity they have and what we've commited to in the run of the task
manager. The other is a DependencyGraph for tracking job dependencies.

Also, use the same method to zero out the graph in the two places we
might want to do it. The main reason we would ever actually zero out the
graph in awx/main/managers.py is in tests where we have not already
initialized the instance group capacity graph.

kdelee commented:
Replacing with https://github.com/ansible/awx/pull/11651/commits
